### PR TITLE
UI: Clustering with Custom Distances and Keyers

### PR DESCRIPTION
**Pull request by zyadtaha**
The second pr in the GSoC project #4301 

Changes proposed in this pull request:
- Replace `<select>` with custom dropdown
- Update test setup to include dropdown menu functionality

Here's what the dropdown menu looks like now:
Screenshot from 2024-06-17 08-03-24

### CODE REVIEW

**Comment by zyadtaha** [comment]
Hi @wetneb,

I updated the functions selector to a dropdown menu. Now, I'm considering changing the method selector too. What do you think about that?

Also, should I complete all the frontend development in one pull request (this one), or would you prefer multiple pull requests for easier reviewing?

I know you're busy with the barcamp, so take your time to reply. I'm not in a rush.

Thank you!

**Comment by wetneb** [comment]
Congrats for getting this done so quickly! Indeed the BarCamp is keeping me quite busy but I will try to have a closer look before our meeting on Thursday.

**Comment by wetneb** [comment]
To give a bit more context to the rest of the community, it would be worth pointing to your proposed design, which will help understand why you are converting a native drop-down select to something else (because it would not make sense to do that in isolation, without wanting to make further changes)

> I updated the functions selector to a dropdown menu. Now, I'm considering changing the method selector too. What do you think about that?

I think we should stick to native widgets as much as possible. If you foresee needing to make similar changes to the method selector it would make sense, but I am not sure why we would need that? What sort of changes do you have in mind?

Before reviewing the code in details, I'd like to list the broad sorts of problems I can think of in this case:
* accessibility might not be as good as the native widget. For instance, I noticed that it's not possible (or I didn't figure out how) to select an option using the keyboard only.
* visual style: it needs to look natural. I think you've done a pretty good job at that (at least on my platform and to my taste) but I wonder if it would be worth re-introducing some sort of downwards arrow like the ones native widgets have (at least on my platform):
image

But I think it doesn't need to look exactly like a native drop-down menu. It's likely not possible to do anyway, given that the rendering likely depends on the operating system / browser in many ways. You could also "own it" and go for a fairly different design, I guess.

File: main/webapp/modules/core/styles/dialogs/clustering-dialog.css [hunk 1]
```diff
@@ -166,3 +166,78 @@ table.clustering-dialog-entry-table a:hover {
...
+}
+
+#keyingDropdownBtn {
+  min-width: 130px;
+  text-align: left;
+  background-color: #ffffff;
+  border: 1px solid #727272;
+}
+
+.keyingFunctionSelector {
```
Comment (wetneb) [inline main/webapp/modules/core/styles/dialogs/clustering-dialog.css:186]: Nitpick: for consistency with the rest of the class names, you could stick to this syntax:
```suggestion
.keying-function-selector {
```
and similarly for other CSS rules in this file

**Comment by ostephens** [comment]
> To give a bit more context to the rest of the community, it would be worth pointing to your proposed design, which will help understand why you are converting a native drop-down select to something else (because it would not make sense to do that in isolation, without wanting to make further changes)

Just to give a +1 to this - I'm very interested in understanding the approach being taken overall here

**Comment by zyadtaha** [comment]
> > To give a bit more context to the rest of the community, it would be worth pointing to your proposed design, which will help understand why you are converting a native drop-down select to something else (because it would not make sense to do that in isolation, without wanting to make further changes)
> 
> Just to give a +1 to this - I'm very interested in understanding the approach being taken overall here

The main reason is that we can't dynamically add a new option to the native dropdown (created with the <select> element).

Also, to move the "Add new function" feature into the dropdown menu which will save a lot of space in the already crowded clustering dialog window.

Here's an example.
Screenshot from 2024-06-21 21-15-10

**Comment by zyadtaha** [comment]
I'm not sure why, but it seems the picture didn't show up in my last comment.image

**Comment by tfmorris** [comment]
> The main reason is that we can't dynamically add a new option to the native dropdown (created with the element).

Can you expand on this? The current drop down menus are constructed dynamically from Javascript, so, on the surface, it seems like the same technique could possibly be used to change them at run time.

**Comment by zyadtaha** [comment]
> > The main reason is that we can't dynamically add a new option to the native dropdown (created with the element).
> 
> Can you expand on this? The current drop down menus are constructed dynamically from Javascript, so, on the surface, it seems like the same technique could possibly be used to change them at run time.

Yes, you are correct. I apologize for the confusion. 

The reason is that a `<select>` element is restricted to containing only `<option>` elements.

Here are the implications:
- It's invalid to add an `<hr>` element inside a `<select>` element to separate options.
- It's not possible to add a delete icon to each option using the native dropdown menu.
- As the number of custom expressions grows, we will require dropdown menus with search bars like this. 
Screenshot from 2024-06-22 02-24-57
However, we cannot place the search bar inside the `<select>` element for the same reason as before. This arrangement would appear odd, with the search bar outside the dropdown menu.
Screenshot from 2024-06-22 02-29-02

**Comment by Abbe98** [comment]
Are you sure you want a dropdown menu at all? Seems like it will be very hard to make it accessible and in line with the current UI given your requirements. Why not a modal with a filterable list instead? Or simply a `<datalist>`?

**Comment by antoine2711** [comment]
> * It's invalid to add an `<hr>` element inside a `<select>` element to separate options.

I don’t understand, @zyadtaha.

This works in Safari and in Chrome:
```
  <select name="papersize">
    <option>Select a paper size</option>
    <hr>
    <option>5.5 × 8.5 in</option>
    <option>8.5 × 11.0 in</option>
    <option>8.5 × 14.0 in</option>
    <option>11.0 × 17.0 in</option>
    <hr>
    <option>A3</option>
    <option>A4</option>
    <option>A5</option>
    <option>A6</option>
    <hr>
    <option>Envelope #10</option>
    <option>Envelope B5</option>
    <option>Envelope C5</option>
    <option>Envelope Monarch</option>
</select>
```

What do you mean?

Regards, Antoine

**Comment by zyadtaha** [comment]
> > * It's invalid to add an `<hr>` element inside a `<select>` element to separate options.
> 
> I don’t understand, @zyadtaha.
> 
> This works in Safari and in Chrome:
> 
> ```
>   <select name="papersize">
>     <option>Select a paper size</option>
>     <hr>
>     <option>5.5 × 8.5 in</option>
>     <option>8.5 × 11.0 in</option>
>     <option>8.5 × 14.0 in</option>
>     <option>11.0 × 17.0 in</option>
>     <hr>
>     <option>A3</option>
>     <option>A4</option>
>     <option>A5</option>
>     <option>A6</option>
>     <hr>
>     <option>Envelope #10</option>
>     <option>Envelope B5</option>
>     <option>Envelope C5</option>
>     <option>Envelope Monarch</option>
> </select>
> ```
> 
> See: https://webkit.org/blog/14933/bringing-back-horizontal-rules-in-select-elements/
> 
> What do you mean?
> 
> Regards, Antoine

Thank you for correcting me. It seems I made a mistake in my search. But we still can't add icons or input elements, can we?

**Comment by zyadtaha** [comment]
> Are you sure you want a dropdown menu at all? Seems like it will be very hard to make it accessible and in line with the current UI given your requirements. Why not a modal with a filterable list instead? Or simply a `<datalist>` element?

What are your thoughts regarding the current options? @wetneb @ostephens @tfmorris @antoine2711

**Comment by wetneb** [comment]
I think it's worth experimenting with other designs! I don't know what a "modal with a filterable list" or a "`<datalist>`" would look like. Perhaps you could try re-designing this selection, with the understanding that people are ready to see bigger UI changes there and you don't need to squeeze in your feature in the exact current layout.

**Comment by ostephens** [comment]
@zyadtaha thanks for the extra information. Can I check I understand the expected workflow? Is the idea:

1. User goes to cluster screen
2. User goes to select keying or distance function in dropdown
3. User finds no suitable function, so clicks "Add new"
4. Some screen or dialog presented where user can define their function (have you given any thought to what this would look like yet?)
5. User creates/saves function
6. User returned to select list with new option added

Have I got that correct?

**Comment by zyadtaha** [comment]
> 4. Some screen or dialog presented where user can define their function (have you given any thought to what this would look like yet?)
>
- We've decided to use the current expression dialog, which has tabs for Preview, History, and Favorites. The History tab lets users see past expressions with option to star. This dialog also allows users to choose their preferred expression language, not limited to GREL.
- When adding a custom distance function, it's important to explain which variables users can use, like value1 and value2. To address this, we've included a default expression in the editor that uses these values (for example, editDistance(value1, value2) as shown in #6613).
> 7. User selects new function (or automatically selected?)
>
Yes, I would go for automatic selection. 

> Have I got that correct?
>
Yes, absolutely.

**Comment by ostephens** [comment]
Thanks @zyadtaha. I guess the question then becomes how would the user generally manage the expressions they had written? It seems like there should be some option to Edit and Delete the expression (possibly duplicate as well?). While I can see the create option working OK as you describe above, I think adding in any additional functions could make the UI very crowded.

This makes me think that at least worth considering having the options for managing the available functions somewhere outside the dropdown menu. Potentially the two approaches could be combined of course, but splitting this out could avoid having this custom dropdown and offer a bit more space. I'd be interested to know how others (both on this thread @antoine2711 @Abbe98 @tfmorris and other users who are maybe on the forum but not on GitHub) might feel about this.

If you've already had these discussions somewhere and decided not to go this way then of course I respect that!

**Comment by wetneb** [comment]
The original discussion is indeed on the forum:
https://forum.openrefine.org/t/user-defined-clustering-project/1528/3?u=antonin_d
In this discussion there was support from Benjamin (not sure about GitHub username) for integrating the delete buttons inside the drop-down.

**Comment by ostephens** [comment]
Thanks @wetneb  - that's very helpful. I don't want to re-hash discussions from the forum here. I do think those designs are missing an option to edit the user defined function which I think could be important - I can add a comment to the forum.

**Comment by tfmorris** [comment]
I think @Abbe98 's point about accessibility is a good one. Also, it's a red flag to me that no one has created a filterable pulldown menu with deletable items that you can reuse. Using a dialog instead will give you a lot more real estate and navigational flexibility to work with when designing the UX.

**Comment by zyadtaha** [comment]
Closing this to make the PR simpler.

**Comment by wetneb** [comment]
There is a new version of this PR at #6705: you are all encouraged to chime in on the new design there.

**Pull request closed without merging**